### ISSUE TITLE
Problem: Modular system with alembic Multiple Bases (like django migrations)

### ISSUE THREAD

**Issue by python-programmer** [sqlalchemy/alembic#777]
## What I want
I want to design a structure like django

Imagine:

I have two apps: polls, profiles

and I want to have this project structure:

```
Project-Root
|
________ settings-package
|
________ polls-app-package
        |
________ ________ migrations
                |
________________________ env.py
________________________ script.py.mako
________________________ versions
|
________ profiles-app-package
        |
________ ________ migrations
                |
________________________ env.py
________________________ script.py.mako
________________________ versions
|
________ alembic.ini

```

this is the alembic.ini files:

```
[alembic]
script_location = migrations

version_locations = %(here)s/polls/migrations/versions %(here)s/profiles/migrations/versions migrations/versions


sqlalchemy.url = postgresql://dev_admin:dev_password@localhost/mdb_db


[post_write_hooks]

[loggers]
keys = root,sqlalchemy,alembic

[handlers]
keys = console

[formatters]
keys = generic

[logger_root]
level = WARN
handlers = console
qualname =

[logger_sqlalchemy]
level = WARN
handlers =
qualname = sqlalchemy.engine

[logger_alembic]
level = INFO
handlers =
qualname = alembic

[handler_console]
class = StreamHandler
args = (sys.stderr,)
level = NOTSET
formatter = generic

[formatter_generic]
format = %(levelname)-5.5s [%(name)s] %(message)s
datefmt = %H:%M:%S


[polls]
script_location = polls/migrations

[profiles]
script_location = profiles/migrations
```

## Process

a normal development process, I create the polls app and create migrations with:

```
alembic --name=polls revision -m "create polls branch" --head=base --branch-label=polls --version-path=polls/migrations/versions --autogenerate
```

and then apply migrations:

```
alembic --name=polls upgrade polls@head 
```

so far, everything is fine

then I want to create another app named profiles

a problem is here:

migrations command:
```
alembic --name=profiles revision -m "create profiles branch" --head=base --branch-label=profiles --version-path=profiles/migrations/versions --autogenerate
```

This gives me an error that makes sense:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
ERROR [alembic.util.messaging] Can't locate revision identified by '1b1810b25190'
  FAILED: Can't locate revision identified by '1b1810b25190'
```

And

**Question**: how to resolve this problem?

**Comment by zzzeek** [comment]
you have two individual environments referring to the same database.  so you have to make sure there are two distinct alembic_version tables.  set this in each env.py using the [version_table](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=version_table#alembic.runtime.environment.EnvironmentContext.configure.params.version_table) parameter.

**Comment by python-programmer** [comment]
thank you, 
I have another issue, 
when I create a migration for polls (version_table = poll_migration), and apply it to the db, 
then I create a migration for profiles, in profiles migrations (version_table = profiles_migration), it deletes polls table (downgrade)

```
"""create profiles branch

Revision ID: 073f840b4ed5
Revises: 
Create Date: 2021-01-16 15:04:29.656971

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '073f840b4ed5'
down_revision = None
branch_labels = ('profiles',)
depends_on = None


def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('profiles',
    sa.Column('id', sa.Integer(), nullable=False),
    sa.Column('first_name', sa.String(length=200), nullable=True),
    sa.Column('last_name', sa.String(length=200), nullable=True),
    sa.Column('birthday', sa.DateTime(), nullable=True),
    sa.PrimaryKeyConstraint('id')
    )
    op.drop_table('polls_migrations')
    # ### end Alembic commands ###


def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('polls_migrations',
    sa.Column('version_num', sa.VARCHAR(length=32), autoincrement=False, nullable=False),
    sa.PrimaryKeyConstraint('version_num', name='polls_migrations_pkc')
    )
    op.drop_table('profiles')
    # ### end Alembic commands ###

```


I don't want to delete the polls table

I want each app migrations to be independent

**Comment by zzzeek** [comment]
> thank you,
> I have another issue,
> when I create a migration for polls (version_table = poll_migration), and apply it to the db,
> then I create a migration for profiles, in profiles migrations (version_table = profiles_migration), it deletes polls table

A nitpick, but an important one - Alembic is not "deleting" your "polls" table.  The autogenerate feature of Alembic is *proposing* that based on what it is observing about your database compared to the MetaData object you are sending, that the "polls" table be dropped when the migration is run.  It places this in a function body that has a comment "commands auto generated by Alembic - please adjust!" at the top.

In this case, you are running autogenerate under two environemnts against the same database, so Alembic has no way to know, when it sees a table in the database not present in your MetaData(), that this table wasn't dropped.

To give Alembic the information it needs to make this decision you must provide an [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) function that provides this special ruleset you are looking for, such as:


```
metadata_one = ...
metadata_two = ...

def include_object(object, name, type_, reflected, compare_to):
   if type_ == "table" and reflected and name not in metadata_two.tables:
       return False

```

**Comment by python-programmer** [comment]
> > thank you,
> > I have another issue,
> > when I create a migration for polls (version_table = poll_migration), and apply it to the db,
> > then I create a migration for profiles, in profiles migrations (version_table = profiles_migration), it deletes polls table
> 
> A nitpick, but an important one - Alembic is not "deleting" your "polls" table. The autogenerate feature of Alembic is _proposing_ that based on what it is observing about your database compared to the MetaData object you are sending, that the "polls" table be dropped when the migration is run. It places this in a function body that has a comment "commands auto generated by Alembic - please adjust!" at the top.
> 
> In this case, you are running autogenerate under two environemnts against the same database, so Alembic has no way to know, when it sees a table in the database not present in your MetaData(), that this table wasn't dropped.
> 
> To give Alembic the information it needs to make this decision you must provide an [include_object](https://alembic.sqlalchemy.org/en/latest/api/runtime.html?highlight=include_object#alembic.runtime.environment.EnvironmentContext.configure.params.include_object) function that provides this special ruleset you are looking for, such as:
> 
> ```
> metadata_one = ...
> metadata_two = ...
> 
> def include_object(object, name, type_, reflected, compare_to):
>    if type_ == "table" and reflected and name not in metadata_two.tables:
>        return False
> ```

thank you @zzzeek ,

I wrote this:

```
# env.py (main file)


def _include_object(target_schema):
    def include_object(obj, name, object_type, reflected, compare_to):
        print(name, object_type)
        if object_type == "table" and reflected and name not in target_schema:
            return False
        return True

    return include_object
```

and this:

```
def run_migrations_offline(configuration: dict, schemas: List[str]):
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """
    context.configure(
        literal_binds=True,
        dialect_opts={"paramstyle": "named"},
        compare_type=True,
        include_schemas=True,
        include_object=_include_object(schemas),
        **configuration
    )

    with context.begin_transaction():
        context.run_migrations()
```

for those table that already exists everything is fine,

but when I added a new app, for an example **products** and ran migrations, the products migrations file was **empty**, and **_include_object** didn't print **products** table (the print statement in the _include_object funciton)

migrations command:

```
alembic --name=products revision -m "products models added" --autogenerate --head=base --branch-label=products
```

and this is output:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.ddl.postgresql] Detected sequence named 'choices_id_seq' as owned by integer column 'choices(id)', assuming SERIAL and omitting

choices table --------> print 
polls_migrations table  --------> print 

INFO  [alembic.ddl.postgresql] Detected sequence named 'profiles_id_seq' as owned by integer column 'profiles(id)', assuming SERIAL and omitting

profiles table  --------> print 
profiles_migrations table  --------> print 
games_migrations table  --------> print 
questions table  --------> print 

INFO  [alembic.ddl.postgresql] Detected sequence named 'games_id_seq' as owned by integer column 'games(id)', assuming SERIAL and omitting

games table  --------> print 
id column  --------> print 
name column  --------> print 
pts column  --------> print 
pub_date column  --------> print 

  Generating /Users/Documents/Projects/Python/apps/products/migrations/versions/4552daa7f6ae_products_models_added.py ...  done
```

when using  the **_include_object**: first migrations does not find new schemas ?

**Comment by zzzeek** [comment]
not totally sure what the question is, so here are some bits of information about new schemas:

include_object gives you visibility into every object that autogenerate sees from the Table level down.  Each table has a .schema attribute on it, so if you are also doing include_schemas, you need to filter on this as well.

There is an issue #650 that proposes that names like schema names, table names etc. also be filterable before they are reflected.   that would make what you are trying to do a little easier, but for now you need to look at the .schema attribute on each Table if you are also being sensitive to that.

if the issue is you have Table objects with a new .schema name and there is no CREATE SCHEMA in the database, Alembic doesn't cover that right now.

**Comment by python-programmer** [comment]
thank you @zzzeek,
schema, schema, schema, I didn't set it correctly

**Comment by python-programmer** [comment]
I'm sorry, I asked a lot of questions

my final question: 

the polls app has a table named questions

the profiles app has a table named profiles

**I have a version_table for each app**

I want to add a foreign key (from profiles table) to questions like this

questions model
```
...
user = Column(Integer, ForeignKey('profiles.id'))

```

it gives me an error:

```
Foreign key associated with column 'questions.user' could not find table 'profiles' with which to generate a foreign key to target column 'id'

```

**Question**: can each table find other app tables?

**Comment by zzzeek** [comment]
cross-metadata foreign keys need to use the column object:

    Table("mytable", metadata, Column("col", ForeignKey(other_table.c.other_col)))

also the include_name hook that will help a lot w/ your use case is coming up in alembic 1.5

**Comment by python-programmer** [comment]
> I'm sorry, I asked a lot of questions
> 
> my final question:
> 
> the polls app has a table named questions
> 
> the profiles app has a table named profiles
> 
> **I have a version_table for each app**
> 
> I want to add a foreign key (from profiles table) to questions like this
> 
> questions model
> 
> ```
> ...
> user = Column(Integer, ForeignKey('profiles.id'))
> ```
> 
> it gives me an error:
> 
> ```
> Foreign key associated with column 'questions.user' could not find table 'profiles' with which to generate a foreign key to target column 'id'
> ```
> 
> **Question**: can each table find other app tables?


Since version 1.5 of this program has been released,
How can this problem be solved?

**Comment by zzzeek** [comment]
> > I'm sorry, I asked a lot of questions
> > my final question:
> > the polls app has a table named questions
> > the profiles app has a table named profiles
> > **I have a version_table for each app**
> > I want to add a foreign key (from profiles table) to questions like this
> > questions model
> > ```
> > ...
> > user = Column(Integer, ForeignKey('profiles.id'))
> > ```
> > 
> > 
> > it gives me an error:
> > ```
> > Foreign key associated with column 'questions.user' could not find table 'profiles' with which to generate a foreign key to target column 'id'
> > ```
> > 
> > 
> > **Question**: can each table find other app tables?
> 
> Since version 1.5 of this program has been released,
> How can this problem be solved?

are you referring to the error "Foreign key associated with column 'questions.user' could not find table 'profiles' with which to generate a foreign key to target column 'id'"?   The answer is as I stated in https://github.com/sqlalchemy/alembic/issues/777#issuecomment-762439856  ; two Table objects belonging to two separate MetaData collections can cross-reference each other using the Column object itself within the ForeignKey definition, not a string name.

Sorry if my mention of the "include_name" hook in that same answer misled you into thinking something regarding this was "fixed" in 1.5.
